### PR TITLE
ENH: add `leaves_color_list` for cluster.dendrogram dictionary.

### LINKED ISSUE #12190: retrieve leave colors from dendrogram
**Is your feature request related to a problem? Please describe.**
I can not retrieve leaves colors from [scipy.cluster.hierarchy.dendrogram](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.dendrogram.html). I need it to color other type of graphics. An example is posted [here](https://stackoverflow.com/questions/61959602/retrieve-leave-colors-from-scipy-dendrogram).

**Describe the solution you'd like**
Include a new key in the dendrogram dictionary informing about color leaves.

**Describe alternatives you've considered**
Provide in the meantime an algorithm to obtain the leaves colors using the information contained in the current dendrogram dictionary, if that's possible. Or another workaround.

**Additional context (e.g. screenshots)**
Code example:
```
import numpy as np
import matplotlib.pyplot as plt
from scipy.cluster.hierarchy import linkage, dendrogram

# DATA EXAMPLE
x = np.array([[ 5, 3],
              [10,15],
              [15,12],
              [24,10],
              [30,30],
              [85,70],
              [71,80]])

# DENDROGRAM
plt.figure()
plt.subplot(121)
z = linkage(x, 'single')
d = dendrogram(z)

# COLORED PLOT
# This is what I would like to achieve. Colors are assigned manually by looking
# at the dendrogram, because I failed to get it from d['color_list'] (it refers 
# to links, not observations)
plt.subplot(122)
points = d['leaves']
colors = ['r','r','g','g','g','g','g']
for point, color in zip(points, colors):
    plt.plot(x[point, 0], x[point, 1], 'o', color=color)
```
Figure output from code example:
image

### PR DESCRIPTION
**Pull request by AtsushiSakai**
#### Reference issue
fix #12190 

#### What does this implement/fix?
I added a new key `leaves_color_list` in the dendrogram dictionary informing about color leaves.
This PR enables for users to retrieve leaves colors from it.

#### Additional information
This is a sample usage:
```
import matplotlib.pyplot as plt
import numpy as np

from scipy.cluster.hierarchy import linkage, dendrogram

x = np.array([[5, 3],
              [10, 15],
              [15, 12],
              [24, 10],
              [30, 30],
              [85, 70],
              [71, 80]])
z = linkage(x, 'single')
d = dendrogram(z)
points = d['leaves']
colors = d['leaves_color_list']
for point, color in zip(points, colors):
    plt.plot(x[point, 0], x[point, 1], 'o', color=color)
plt.show()
```
SS 2020-06-30 at 22 18 33

### CODE REVIEW

File: scipy/cluster/hierarchy.py [hunk 1]
```diff
@@ -3358,9 +3362,23 @@ def llf(id):
...
 
 
+def _get_leaves_color_list(R):
+    leaves_color_list = [None] * len(R['leaves'])
+    for link_x, link_y, link_color in zip(R['icoord'],
+                                          R['dcoord'],
+                                          R['color_list']):
+        for (xi, yi) in zip(link_x, link_y):
+            if yi == 0.0:  # if yi is 0.0, the point is a leaf
+                # xi is 5, 15, 25, 35, ...
```
Comment (rgommers) [inline scipy/cluster/hierarchy.py:3377]: This confused me. It seems to be related to the hardcoded x-axis values in `iv_ticks`. Could you expand the comment a bit to make clear what's going on here?

Comment (AtsushiSakai) [reply scipy/cluster/hierarchy.py]: Thank you for reviewing. I added additional comments. PTAL

**Comment by rgommers** [comment]
Looks like this works as advertised. I'm still slightly worried about that counting logic so would like to test on some real-world data. Let's ping @alejandro-ariza who requested this feature. @alejandro-ariza can you please test this implementation?

**Comment by rgommers** [comment]
There's also some bio data at https://github.com/scipy/scipy/issues/5238#issuecomment-249499573 that's a nice test, with some more colors.

**Comment by AtsushiSakai** [comment]
> There's also some bio data at #5238 (comment) that's a nice test, with some more colors.

I tried this nice test.
This is the code:
```
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy.cluster.hierarchy import linkage, dendrogram

gasch_data = 'experiment_data.txt'
dists = pd.io.parsers.read_table(gasch_data, index_col=0)
dists = np.asarray(dists)
Z = linkage(dists, 'ward')
plt.figure()
dn = dendrogram(Z, truncate_mode='lastp',
                p=40, show_contracted=True,
                orientation='right')
labels = np.array(dn['ivl'])
colors = np.array(dn['leaves_color_list'])

for color in np.unique(colors):
    print(color, ":", labels[np.where(colors == color)])
```

This is the dendrogram figure
dendrogram_test

This is the output of the code:
```
C1 : ['(39)' '(112)' '(97)' '(92)' '(92)' '(41)']
C2 : ['(104)' '(58)' '(11)' '1356' '(22)' '(19)' '(44)']
C3 : ['(37)' '(17)' '(38)' '(233)' '(102)' '(90)' '(80)' '(291)' '(14)' '(146)'
 '(146)' '(145)' '(121)' '(100)' '(130)' '(262)' '(244)' '(16)' '(280)'
 '(220)' '(178)' '(141)' '(72)' '(72)' '(317)' '(460)' '(83)']
```

The code looks work correctly.

**Review by rgommers — APPROVED** [review]
Great, thanks @AtsushiSakai! Merging.

**Comment by spackows** [comment]
Hi.  I'm trying to use `leaves_color_list` to color the labels in my dendrogram plot.

### Code
```
import numpy as np
import matplotlib.pyplot as plt
from scipy.cluster.hierarchy import linkage, dendrogram

# Data
x = np.array([[1, 0, 0], 
              [0, 0, 1], 
              [0, 2, 0], 
              [0, 0, 1], 
              [0, 1, 0], 
              [0, 1, 0]] )

# DENDROGRAM
plt.figure( figsize=( 15, 3 ) )
plt.box( False )
z = linkage( x, "single" )
d = dendrogram( z )

for label, leaf_color in zip( plt.gca().get_xmajorticklabels(), d["leaves_color_list"] ):
    label.set_color( leaf_color )

print( d["leaves_color_list"] )
```

### Output

image

### Problem

You can see that the left-most label is not the right color in a few cases: 4 and 1 .

Is the nature of my data unsupported for creating a dendrogram?   (My matrices are normally much larger and the non-zero values are usually low-value integers.  But I wanted to create a simple example for the repro here.)

Or might there be a bug in the generation of the `leaves_color_list`?  

If it is a bug, can you tell me where to open a new issue so this can be fixed?

Thanks!

**Comment by AtsushiSakai** [comment]
@spackows Thank you for reporting. I'm not sure the behavior is expected or not, but the doc of `dendrogram` says:
https://github.com/scipy/scipy/blob/4cf21e753cf937d1c6c2d2a0e372fbc1dbbeea81/scipy/cluster/hierarchy.py#L3253-L3254
I think your x[: ,2] is not monotonic
> [0 1 0 1 0 0]

Is it expected?

**Comment by spackows** [comment]
Hi.

Thank you for your reply.

<p>&nbsp;</p>

> It is expected that the distances in ``Z[:,2]`` be monotonic ...

Is it `x[ :, 2 ]` - the input matrix of data passed to `linkage` - that needs to be monotonic?

Or is it `z[ :, 2]` - the result from `linkage`, the distance matrix passed to `dendrogram` - that needs to be monotonic?

The docs _say_ "Z".. which makes me think they mean the later.

When I run this code:

```
x = np.array([[1, 0, 0], 
              [0, 0, 1], 
              [0, 2, 0], 
              [0, 0, 1], 
              [0, 1, 0], 
              [0, 1, 0]] )

z = linkage( x, "single" )

print( z[ :, 2 ] )
```

I get this output, which is monotonic:

```
[ 0.         0.         1.         1.41421356 1.41421356 ]
```

<p>&nbsp;</p>

> ... otherwise  crossings appear in the dendrogram. 

The problem with `leaves_color_list` isn't that it's out of order (eg. what I might expect to happen if the lines crossed.)  

<p>&nbsp;</p>

## Hypothesis

It looks like an edge case where the distance is zero.

When I change one value in the input data so the distances is not quite zero, the leaf color is correct.

**Code**

```
x = np.array([[1, 0, 0], 
              [0, 0, 1.1], 
              [0, 2, 0], 
              [0, 0, 1], 
              [0, 1, 0], 
              [0, 1, 0]] )

plt.figure( figsize=( 15, 3 ) )
plt.box( False )
z = linkage( x, "single" )
d = dendrogram( z )

for label, leaf_color in zip( plt.gca().get_xmajorticklabels(), d["leaves_color_list"] ):
    label.set_color( leaf_color )

print( d["leaves_color_list"] )
```

**Output**

image

**Comment by AtsushiSakai** [comment]
Ah, sorry. I misunderstood. You are right.
Could you please create an issue for your problem?
I will try to check it deeply.

**Comment by spackows** [comment]
Thank you for your attention to this discussion.  

I opened this issue to investigate the zero-distance problem: https://github.com/scipy/scipy/issues/16558

\*I have never contributed to **scipy** before.  But I would be happy to help with this problem, if I can - such as by trying workarounds or testing a fix.

Thank you again!

**Pull request merged**